### ISSUE TITLE
Update Citrine to Ubuntu 22.04 performance issues

### ISSUE THREAD

**Issue by franz1981** [TechEmpower/FrameworkBenchmarks#8038]
As commented by @nbrady-techempower in https://github.com/TechEmpower/FrameworkBenchmarks/issues/7321#issuecomment-1466365129 here we are!

I don't have a clue why/what's going on, but let's take a look at both vertx and redkale before/after the upgrade, with **Plaintext**:

[before](https://www.techempower.com/benchmarks/#section=test&runid=ef7991e0-35e3-4de1-94b6-1ac13e29b9c6&test=plaintext&f=zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-xan9xb-zik0zj-zik0zj-z8kflr-zik0zj-9zldr):
![image](https://user-images.githubusercontent.com/13125299/225314979-d4d3b4d5-c861-40a4-8d07-9c700272c912.png)

and @redkale, [last run](https://www.techempower.com/benchmarks/#section=test&runid=9072dc54-398e-4e79-b4bc-0d0a51c5aa8d&test=plaintext&f=zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-xan9xb-zik0zj-zik0zj-z8kflr-zik0zj-jz6rj):
![image](https://user-images.githubusercontent.com/13125299/225315271-385139e1-1cac-46e2-9a2b-be5684213ab0.png)

and both vert.x and redkale in [another nightly](https://www.techempower.com/benchmarks/#section=test&runid=35d492e1-30d0-44ad-b21a-e6d9ecb14c06&f=zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-zik0zj-xan9xb-zik0zj-zik0zj-z8kflr-zik0zj-9zldr&test=plaintext) with the new env:
![image](https://user-images.githubusercontent.com/13125299/225315775-2e097662-68d7-4403-b4f9-a0f284f1858a.png)

It appears clear to me that the new env is quite unstable for some top performers and given the discussion at https://github.com/TechEmpower/FrameworkBenchmarks/discussions/7984#discussioncomment-5282261 where RedKale isn't performing any http request header decoding and hasn't changed version across the different runs, something wrong is going on here...

**Comment by nbrady-techempower** [comment]
I'm seeing some pretty good consistency between new runs on the new environment for the top performers and spot checking some others. I'm not yet convinced it's the environment that's unstable. 

I believe Redkale was the framework I saw that was changing snapshots between runs so that they didn't have to alter the version number in our repo, so frankly, I don't trust the variance we're seeing here from them.

**Comment by franz1981** [comment]
And what about Vertx and Netty or Helidon Nima?
These are all JVM based ones and none of them has upgraded or changed anything before/after the env change.

**Comment by nbrady-techempower** [comment]
Netty and Helidon Nima both are consistent across runs in the new env. So, I don't believe the new environment is "unstable." However, it could be that some part of the update has affected the performance of some frameworks that could have been uniquely tuned to the previous environment. Do you have any insight into what that might be?

**Comment by franz1981** [comment]
Sorry, with Netty I was wrong, but Helidon Nima has received an hit and Vertx the same. They didn't have any specific optimization at socket/env or OS level - no affinity (I was involved into the vertx one given that I am responsible of it with Julien Viet).
Nima and Vertx are using too distant JVM versions and they are using very different Epoll optimizations under the hood, meaning that anything related networking and JVM can be excluded.

The sole thing that make them similar is that they are CPU bottlenecked (and can be observed in the dstat metrics), meaning that even a small noise there can cause perf variations.

**Comment by nbrady-techempower** [comment]
>  but Helidon Nima has received an hit

But you mean from the old environment to the new environment? Because in two runs on the new environment, helidon nima is 3,584,007 and 3,556,068 which is normal variance as far as I'm concerned.

**Comment by franz1981** [comment]
Yep, exactly, and for vertx the same - the old env vs new env numbers.
I am still waiting vertx current nightly to see if the last new env result will repeat. 
Sorry if I wasn't clear enough to explain that was an old vs new env perf hit.

**Comment by nbrady-techempower** [comment]
Ok, yes. I agree with this. It seems like there's been a small hit (some large ones) across the board for a lot of frameworks fairly consistently. I'm happy to relay some information from the environments to help people track down what that might be.

**Comment by franz1981** [comment]
Many thanks! The only thing that come in my mind related both Vert-x and Nima (built on 2 different stacks) is that they are both CPU limited and are overcommiting - the former by using twice the number of event loops vs number of available cores, the other by using a rightly sized Fork Join thread pool  but some additional competing and busy garbage collection threads.
It makes me think that something related IRQ processing or anything else that can steal CPU cycles can hit both (and the perf hit looks similar too - ~8/9%)

**Comment by franz1981** [comment]
Let me think if adding some profiling could be of any help to see what's going on or even checking how the CPU usage of top performers has changed if they were not running at 100% CPU before - and looking at https://ajdust.github.io/tfbvis/?testrun=Citrine_started2022-06-30_edd8ab2e-018b-4041-92ce-03e5317d35ea&testtype=plaintext&round=false I see that they were not maxing out CPU

**Comment by redkale** [comment]
@franz1981    The reason for the poor performance of the redkale framework this time is that http header coding is enabled, not because of the change of the environment.
   HttpContent.lazyHeaders is an automatic optimization highlight. Redkale will load all RestServices and HttpServlets when the service is started. If all RestServices do not want to read cookie, sessionid, and http header parameters, and there are no HttpFilter or WebSocket, then lazyHeaders=true, http headers will not be parsed. Otherwise, http headers will be parsed。
   My English is poor. The above is translated by robots. What sentences are there。

**Comment by franz1981** [comment]
Thanks @redkale for the comment. I see instead that graal native is performing better then JVM so maybe there the lazy loading is stil happening?

**Comment by franz1981** [comment]
@nbrady-techempower 
Let's try first to exclude the obvious reasons:
- Netty is being configured to use io_uring if the kernel support it
- Vert-x not yet

There is a chance that Vert-x is regressed as many others by the mentioned percentage, but Netty isn't because it has balanced its regression with the improvement provided by io_uring (hence both would have regressed if they use Epoll).
Netty should print which driver is using, can you grab the output of the server while running on citrine?

**Comment by fakeshadow** [comment]
> can you grab the output of the server while running on citrine?

For completed continuous run you can check the detail where it shows you link for result where log is contained.
for example this was from the last run:
https://tfb-status.techempower.com/unzip/results.2023-03-16-21-13-17-497.zip/results/20230310141027/netty/run/netty.log

**Comment by franz1981** [comment]
Many thanks @fakeshadow didn't know that so...
Half mistery solved @nbrady-techempower 
We do have a regression in both Vert-x and Netty  but Netty didn't seems affected because it is using io_uring under the hood now, and amortized the perf drop...

**Comment by NinoFloris** [comment]
I'm seeing strange things happening for fortunes as well. drogon and asp.net core consistently had about 30% more rps before the upgrade.

**Comment by franz1981** [comment]
Thanks @NinoFloris I am working with my team to provide some regression analysis across latest nightly for all frameworks (will work on it next week) so we can help @nbrady-techempower to detect how many have regressed and by how much (and if there are relevant changes between the nightly)

**Comment by volyrique** [comment]
If you check the top performers, it is apparent that there is a performance regression in all tests except cached queries and plaintext - the latter two still being bottlenecked by the 10 Gb/s network speed. The regression is 5-7% for the JSON serialization, multiple queries, and database updates tests, and 14-17% for the single query and fortunes ones; overall, a 9% reduction in composite score.

**Comment by franz1981** [comment]
@redkale Looking the last results it appears you are not performing http request headers decoding anymore again.

**Comment by redkale** [comment]
> @redkale Looking the last results it appears you are not performing http request headers decoding anymore again.

The last time http header parsing was enabled, it was for comparison. For simple http, it is disabled by default

**Comment by synopse** [comment]
@redkale As far as I can understand TFB requirement number 2, the HTTP should be minimal, but realistic.
In respect to this requirement, any framework not parsing the HTTP headers should be marked as "Stripped".

For instance, any minimal HTTP server should properly react to a `Connection: close` header, and actually close the connection.
See https://github.com/TechEmpower/FrameworkBenchmarks/issues/8205

**Comment by NateBrady23** [comment]
Closing as part of bulk cleanup.
